### ISSUE TITLE
Unscoped Crawl-delay is treated as parse error, and thus disallows all URLs

### ISSUE THREAD

**Issue by ivan-is** [seomoz/reppy#47]
Wrong answer whether a URL is allowed with such robots rules:

```
Crawl-delay: 1

User-agent: *
Disallow: /*%
Disallow: /*.aspx
Disallow: /*.aspx*
```
There are a lot of urls with `Crawl-delay:` in the head of the file.

Source url:  `https://learningnetwork.cisco.com/robots.txt`

Steps for reproducing:

```
In [1]: from reppy.cache import RobotsCache
   ...: cache = RobotsCache(capacity=100)

In [16]: cache.allowed('http://learningnetwork.cisco.com', 'my')
Out[16]: False
```

Full robots file:
```
# Puppet managed file

Crawl-delay: 1

Sitemap: https://learningnetwork.cisco.com/sitemap.xml

User-agent: *
Disallow: /*%
Disallow: /*.aspx
Disallow: /*.aspx*
Disallow: /*.jspa
Disallow: /*.jspa*
Disallow: /*/authors/
Disallow: /*/bookmarks
Disallow: /*/customize-container.jspa
Disallow: /*secondVersionNumber*
Disallow: /*/profile-status-list.jspa
Disallow: /*/render-widget!execute.jspa
Disallow: /*/restore
Disallow: /*/restore?version=
Disallow: /*/spotlight-search.jspa
Disallow: /*/status
Disallow: /*/version
Disallow: /*?author
Disallow: /*?containerType=
Disallow: /*@
Disallow: /*=tags
Disallow: /*answerCategory=
Disallow: /*author=
Disallow: /*bookmarks*
Disallow: /*filterID=
Disallow: /*itemView=
Disallow: /*jsessionid=
Disallow: /*-picker.jspa?
Disallow: /*searchQuestions=
Disallow: /*start=
Disallow: /*tab=
Disallow: /*tags=
Disallow: /*tagSet=
Disallow: /*targetUser=
Disallow: /*tstart=
Disallow: /*version=
Disallow: /*view=
Disallow: /bookmarks/
Disallow: /c-
Disallow: /*c-*
Disallow: /docs/*/c-
Disallow: /docs/c-
Disallow: /ITDIT/
Disallow: /mailto*
Disallow: /message/
Disallow: /people/*
Disallow: /poll-post!input.jspa?
Disallow: /post!
Disallow: /post!imagePicker.jspa
Disallow: /post!reply.jspa?messageID=
Disallow: /Products/*.aspx
Disallow: /search.jspa?
Disallow: /servlet/JiveServlet
Disallow: /*servlet*
Disallow: /tags
Disallow: /thread/*/c-
Disallow: /thread/c-
Disallow: /thread-to-doc.jspa
Disallow: /user-picker.jspa
```

**Comment by b4hand** [comment]
Can you please include what version of reppy you're using? In `rep-cpp`, I see an exception thrown with a message saying `"Need User-Agent"` which is fairly reasonable behavior for a malformed robots.txt. Yes, I realize that you may be seeing this as a common error, but I don't think it's reasonable behavior for reppy to assume a malformed robots.txt is allowing robots.

Similarly, using reppy directly, I see the following:

```
>>> robot = Robots.parse('http://learningnetwork.cisco.com/robots.txt', """
... Crawl-delay: 1
...
... User-agent: *
... Disallow: /*%
... Disallow: /*.aspx
... Disallow: /*.aspx*
... """)
Traceback (most recent call last):
  File "<stdin>", line 8, in <module>
  File "reppy/robots.pyx", line 71, in reppy.robots.ParseMethod (reppy/robots.cpp:2375)
  File "reppy/robots.pyx", line 129, in reppy.robots.Robots.__init__ (reppy/robots.cpp:3947)
ValueError: Need User-Agent
```

You may be able to handle this by changing the Cache policy to use something *other* than `ReraiseExceptionPolicy`.

**Comment by ivan-is** [comment]
Hi.
Reppy version:
```
pip freeze | grep reppy
reppy==0.4.3
```

Regarding DEFAULT_CACHE_POLICY:

```
class RobotsCache(BaseCache):
    """A cache of Robots objects."""

    DEFAULT_CACHE_POLICY = DefaultObjectPolicy(ttl=TTL, factory=AllowAll)
```
Why do you think this robots.txt is malformed?

**Comment by b4hand** [comment]
The `Crawl-Delay` declaration *must* appear under the scope of `User-Agent` declaration. It is technically not allowed at the "top-level". The correct order would be:

```
User-agent: *
Crawl-delay: 1
Disallow: /*%
Disallow: /*.aspx
Disallow: /*.aspx*
```

The original given example is unclear which bot the `Crawl-delay` is intended to apply to. There is no implicit wildcard scope for `Crawl-delay`. You must declare `User-agent: *` first if you want it to apply globally according to the specification.

**Comment by dlecocq** [comment]
I've been sort of on the fence about how to deal with this (directives appearing outside of a `User-Agent` block). On the one hand, it is malformed, per the RFC (in the case of `Allow` and `Disallow`) and presumably by extension for `Crawl-Delay`. On the other hand, surely some level of flexibility is called for. I see no obvious winner among:

- throw an exception saying it's malformed and why
- ignore those directives that appear outside of `User-Agent` (provided they ought to be in such a block)
- treat those directives as associated with the default `User-Agent` (`*`)

This is actually more common that I'd originally hope. I'm trying to dig up some numbers from some experiments I did.

**Comment by dlecocq** [comment]
In a sample of about 7.7M URLs taken from Silo (a Moz-internal source of URLs that are believed to be on relatively high-quality sites):

```
7249712 (93.58%) allowed
 497216 ( 6.42%) disallowed (via `Disallow` or exception)
 144112 ( 1.86%) exceptions
```

Just under 2% experienced an exception while trying to check if a page was allowed. The exceptions encountered:

```
Exceptions by type
------------------
  77720 ( 1.00%) ValueError
  23192 ( 0.30%) ConnectionException
  20656 ( 0.27%) TimeoutException
  18603 ( 0.24%) BadStatusCode
   2824 ( 0.04%) ExcessiveRedirects
    658          SSLException
    137          DecodeError
    116          ContentTooLong
    114          MalformedUrl
     86          ProtocolError
      6          LocationValueError
```

Some of these are transient or at worst, unrelated to `reppy` itself. But for 1% of URLs that got a `ValueError`:

```
ValueError
----------
  69071 ValueError('Need User-Agent',)     # Starts with "Disallow"
   8608 ValueError('Need User-Agentm',)    # Starts with "Allow"
     41 ValueError('stoi',)
```

The `stoi` errors come from parsing this (notice the `Allow` directive has an odd URL. But this is small potatoes, and an issue for `url-py` and `url-cpp`.

```
User-Agent: *
Allow: //sitemap:http://www.lifeguardbreathtester.com/sitemap.xml

User-Agent:*
Disallow: /staging/
```

More interesting is the fact that essentially about 1% of URLs have a `robots.txt` that experiences this issue. Not a disastrous portion, but enough to be problematic for a library with high aims.

Methodological note: these URLs were sampled directly from Silo, with no uniquification. So necessarily, this describes fewer than 7.7M FQDNs (and thus fewer than 7.7M `robots.txt` files).

**Comment by b4hand** [comment]
I think choice 3 is reasonable; especially, if you create a new top-level User-Agent context that is distinct from the literal `*` agent. This way you could actually report on them like site issues for Aardwolf, but still handle them gracefully.

**Comment by ivan-is** [comment]
agree with point 3, it would be more preferable

**Comment by b4hand** [comment]
This was resolved by #54. Unscoped Crawl-Delay directives are treated as the `*` agent.
